File: Algorithms/Python/tim_sort.py
from __future__ import annotations


def insertion_sort(array: list[int]) -> list[int]:
    """Sort an array using the insertion sort algorithm."""
    for i in range(1, len(array)):
        key = array[i]
        j = i - 1
        while j >= 0 and key < array[j]:
            array[j + 1] = array[j]
            j -= 1
        array[j + 1] = key
    return array


def merge(left: list[int], right: list[int]) -> list[int]:
    """Merge two sorted arrays."""
    result = []
    i = j = 0
    while i < len(left) and j < len(right):
        if left[i] < right[j]:
            result.append(left[i])
            i += 1
        else:
            result.append(right[j])
            j += 1
    result.extend(left[i:])
    result.extend(right[j:])
    return result


def tim_sort(array: list[int], run: int = 32) -> list[int]:
    """Sort an array using the Tim sort algorithm."""
    for i in range(0, len(array), run):
        array[i : i + run] = insertion_sort(array[i : i + run])
    size = run
    while size < len(array):
        for left in range(0, len(array), 2 * size):
            mid = left + size - 1
            right = min(left + 2 * size - 1, len(array) - 1)
            array[left : right + 1] = merge(array[left : mid + 1], array[mid + 1 : right + 1])
        size *= 2
    return array

File: Algorithms/C++/tim_sort.cpp
#include <iostream>
#include <vector>

void insertion_sort(std::vector<int> &arr)
{
    int size = (int)arr.size();
    for (int i = 1; i < size; i++)
    {
        int key = arr[i];
        int j = i - 1;

        while (j >= 0 && arr[j] > key)
        {
            arr[j + 1] = arr[j];
            j--;
        }

        arr[j + 1] = key;
    }
}

void print_array(const std::vector<int> &arr)
{
    for (int i : arr)
    {
        std::cout << i << " ";
    }

    std::cout << std::endl;
}

void merge(std::vector<int> &arr, int l, int m, int r)
{
    int n1 = m - l + 1;
    int n2 = r - m;

    std::vector<int> L(n1);
    std::vector<int> R(n2);

    for (int i = 0; i < n1; i++)
    {
        L[i] = arr[l + i];
    }

    for (int i = 0; i < n2; i++)
    {
        R[i] = arr[m + 1 + i];
    }

    int i = 0;
    int j = 0;
    int k = l;

    while (i < n1 && j < n2)
    {
        if (L[i] <= R[j])
        {
            arr[k] = L[i];
            i++;
        }
        else
        {
            arr[k] = R[j];
            j++;
        }
        k++;
    }

    while (i < n1)
    {
        arr[k] = L[i];
        i++;
        k++;
    }

    while (j < n2)
    {
        arr[k] = R[j];
        j++;
        k++;
    }
}

void tim_sort(std::vector<int> &arr)
{
    const int RUN = 32;
    int n = (int)arr.size();

    for (int i = 0; i < n; i += RUN)
    {
        insertion_sort(arr);
    }

    for (int size = RUN; size < n; size = 2 * size)
    {
        for (int left = 0; left < n; left += 2 * size)
        {
            int mid = left + size - 1;
            int right = std::min((left + 2 * size - 1), (n - 1));
            merge(arr, left, mid, right);
        }
    }
}

int main()
{
    std::vector<int> arr = {5, 21, 7, 23, 19, 3, 8, 12, 14, 2};
    tim_sort(arr);
    print_array(arr);

    return 0;
}
